devicetypes/widomsrl/widom-smart-dry-contact.src/widom-smart-dry-contact.groovy

```diff
@@ -0,0 +1,548 @@
+    syncNext()
+}
+
+/**
+ *  parseAssocGroupInput(string, maxNodes)
+ *
+ *  Converts a comma-delimited string of destinations (nodes and endpoints) into an array suitable for passing to
+ *  multiChannelAssociationSet(). All numbers are interpreted as hexadecimal. Anything that's not a valid node or
+ *  endpoint is discarded (warn). If the list has more than maxNodes, the extras are discarded (warn).
+ *
+ *  Example input strings:
+ *    "9,A1"      = Nodes: 9 & 161 (no multi-channel endpoints)            => Output: [9, 161]
+ *    "7,8:1,8:2" = Nodes: 7, Endpoints: Node8:endpoint1 & node8:endpoint2 => Output: [7, 0, 8, 1, 8, 2]
+ */
+private parseAssocGroupInput(string, maxNodes) {
+    logging("parseAssocGroupInput(): Parsing Association Group Nodes: ${string}","trace")
+
+    // First split into nodes and endpoints. Count valid entries as we go.
+    if (string) {
+        def nodeList = string.split(',')
+        def nodes = []
+        def endpoints = []
+        def count = 0
+
+        nodeList = nodeList.each { node ->
+            node = node.trim()
+            if ( count >= maxNodes) {
+                logging("parseAssocGroupInput(): Number of nodes and endpoints is greater than ${maxNodes}! The following node was discarded: ${node}","warn")
+            }
+            else if (node.matches("\\p{XDigit}+")) { // There's only hexadecimal digits = nodeId
+                def nodeId = Integer.parseInt(node,16)  // Parse as hex
+                if ( (nodeId > 0) & (nodeId < 256) ) { // It's a valid nodeId
+                    nodes << nodeId
+                    count++
+                }
+                else {
+                    logging("parseAssocGroupInput(): Invalid nodeId: ${node}","warn")
+                }
+            }
+            else if (node.matches("\\p{XDigit}+:\\p{XDigit}+")) { // endpoint e.g. "0A:2"
+                def endpoint = node.split(":")
+                def nodeId = Integer.parseInt(endpoint[0],16) // Parse as hex
+                def endpointId = Integer.parseInt(endpoint[1],16) // Parse as hex
+                if ( (nodeId > 0) & (nodeId < 256) & (endpointId > 0) & (endpointId < 256) ) { // It's a valid endpoint
+                    endpoints.addAll([nodeId,endpointId])
+                    count++
+                }
+                else {
+                    logging("parseAssocGroupInput(): Invalid endpoint: ${node}","warn")
+                }
+            }
+            else {
+                logging("parseAssocGroupInput(): Invalid nodeId: ${node}","warn")
+            }
+        }
+
+        return (endpoints) ? nodes + [0] + endpoints : nodes
+    }
+    else {
+        return []
+    }
+}
+
+
+/*
+##########################
+## Device Configuration ##
+##########################
+*/
+private Map cmdVersions() {
+    [0x5E: 2, 0x86: 2, 0x72: 2, 0x59: 2, 0x98: 1, 0x25: 1, 0x5A: 1, 0x85: 2, 0x70: 1, 0x8E: 2, 0x9F: 1, 0x6C: 1] 
+}
+
+private parameterMap() {[
```

Comment:
Are all those parameters related to scenes and automations necessary here ?

Reply:
The parameters are necessary for change the device behavior(for example number of click on external button in order to turn ON/OFF the load).